src/libpgmoneta/se_ssh.c

```diff
@@ -382,85 +477,640 @@ sftp_copy_directory(char* from, char* to)
+   int read_bytes;
+   char* fn;
+   sftp_file file = NULL;
+   struct backup* bck;
+
+   *backup = NULL;
+
+   fn = NULL;
+   fn = pgmoneta_append(fn, directory);
+   fn = pgmoneta_append(fn, "/");
+   fn = pgmoneta_append(fn, label);
+   fn = pgmoneta_append(fn, "/backup.info");
+
+   file = sftp_open(sftp, fn, O_RDONLY, 0);
+
+   if (file == NULL)
+   {
+      goto error;
+   }
+
+   bck = (struct backup*)malloc(sizeof(struct backup));
+   memset(bck, 0, sizeof(struct backup));
+
+   memcpy(&bck->label[0], label, strlen(label));
+   bck->valid = VALID_UNKNOWN;
+
+   read_bytes = sftp_read(file, &buffer[0], sizeof(buffer));
+
+   if (read_bytes != 0)
+   {
+      char* ptr = NULL;
+      ptr = strtok(&buffer[0], "=\n");
+
+      while (ptr != NULL)
+      {
+         char key[MISC_LENGTH];
+         char value[MISC_LENGTH];
+
+         memset(&key[0], 0, sizeof(key));
+         memset(&value[0], 0, sizeof(value));
+
+         memcpy(&key[0], ptr, strlen(ptr));
+
+         ptr = strtok(NULL, "=\n");
+         memcpy(&value[0], ptr, strlen(ptr));
+
+         if (!strcmp(INFO_STATUS, &key[0]))
+         {
+            if (!strcmp("1", &value[0]))
+            {
+               bck->valid = VALID_TRUE;
+            }
+            else
+            {
+               bck->valid = VALID_FALSE;
+            }
+         }
+         else if (!strcmp(INFO_LABEL, &key[0]))
+         {
+            memcpy(&bck->label[0], &value[0], strlen(&value[0]));
+         }
+         else if (!strcmp(INFO_WAL, &key[0]))
+         {
+            memcpy(&bck->wal[0], &value[0], strlen(&value[0]));
+         }
+         else if (!strcmp(INFO_BACKUP, &key[0]))
+         {
+            bck->backup_size = strtoul(&value[0], &ptr, 10);
+         }
+         else if (!strcmp(INFO_RESTORE, &key[0]))
+         {
+            bck->restore_size = strtoul(&value[0], &ptr, 10);
+         }
+         else if (!strcmp(INFO_ELAPSED, &key[0]))
+         {
+            bck->elapsed_time = atoi(&value[0]);
+         }
+         else if (!strcmp(INFO_VERSION, &key[0]))
+         {
+            bck->version = atoi(&value[0]);
+         }
+         else if (!strcmp(INFO_KEEP, &key[0]))
+         {
+            bck->keep = atoi(&value[0]) == 1 ? true : false;
+         }
+
+         ptr = strtok(NULL, "=\n");
+      }
+   }
+
+   *backup = bck;
+
+   sftp_close(file);
+   free(fn);
+
+   return 0;
+
+error:
+
+   free(fn);
+
+   return 1;
+}
+
+static int
+sftp_write_backup_sha256(char* path, struct hashmap* hash_map)
+{
+   char* buf;
+   char** keys = NULL;
+   char* value = NULL;
+   unsigned int map_size = 0;
+   sftp_file file;
+
+   file = sftp_open(sftp, path, O_WRONLY | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
+
+   if (file == NULL)
+   {
       return 1;
    }
 
-   memset(buffer, 0, sizeof(buffer));
+   map_size = pgmoneta_hashmap_size(hash_map);
+
+   if (pgmoneta_hashmap_key_set(hash_map, &keys))
+   {
+      goto error;
+   }
+
+   for (int i = 0; i < map_size; i++)
```

Comment:
Can't we just write the entire file ?

Reply:
We can directly write into the remote `backup.sha256`. However, for each file we will open the remote file and then append it a new entry because files are copied/linked in `sftp_copy_file()`. This is why I find it better to store all file paths with its corresponding hashes first in a hashmap and then write into `backup.sha256` at the end. What do you think ?

Reply:
Sure, but we need `backup.sha256` locally as well, right ?

Reply:
We can write `backup.sha256` locally at the beginning, and once all files are copied/linked to the remote server. We can copy the local `backup.sha256` to the remote server, and delete the file locally. As the data directory is removed after the remote backup finished.
At this point, whenever `sftp_copy_file()` is triggered, I create a new entry of the file path with its hash and store it in a hashmap. After the remote backup finishes, I write the whole hashmap into the remote `backup.sha256` without the need of having a `backup.sha256` locally.

Reply:
But, `backup.sha256` lives outside of `data/` - next to `backup.info`. So, I'm missing why we need a SHA256 of `backup.sha256` itself

Reply:
There is 2 approaches we can follow one of them:
1- Creating `backup.sha256` locally, and whenever `sftp_copy_file()` is triggered, The file path with its hash will be written into `backup.sha256` locally. At the end of `ssh_storage_execute()` we will copy the `backup.sha256` to the remote server and then we delete the file locally in `ssh_storage_teardown()` as we do the backup/data directory.

2- Storing each file path with its hash in a hashmap and after the remote backup finishes, we iterate over the hashmap and write the entry in `backup.sha256` remotely.

If there is a better approach, you can share it with me.

Reply:
Separate workflow with just the SHA-256 stuff after compression, but before SFTP.

Reply:
https://github.com/pgmoneta/pgmoneta/blob/main/src/libpgmoneta/workflow.c#L133

Reply:
I get it. We will create `backup.sha256` locally in a separate workflow. Is it right ?
If it is, `backup.sha256` stores the remote file path with its hash so that we can just make a link to the latest remote file.
But if we create `backup.sha256` locally before SFTP, we will store the local file path not the remote one.

Reply:
Oh I messed up. I am actually storing the reference path not the full path. You are right. I will create a separate workflow for it.

Reply:
Can't we just store the relative path from `data/` ? We know if we are operating locally or remotely so we can build the full paths from that

Reply:
So,

```
SHA256............:postgresql.conf
```

f.ex

Reply:
Or the other way around

Reply:
Great, lets see if that works better :)

Reply:
I get it.  we will create a separate workflow for writing `backup.sha256` locally before the SFTP, and at the end we will copy this file to the remote server.
In case of reading `backup.sha256` of the latest backup to decide if we either make a copy or do a link. We will read this file locally too ?

Reply:
Well, we can if decide to keep it locally (initial commit I would say yes), otherwise it is a matter of just reading `x` and `x - 1` from the remote server. I think you will get a better idea of what is best during the refactoring

Reply:
Hashmaps will be used only when reading `backup.sha256` ?
We will read it at the beginning and store its entries in hashmap so that we can use it later to decide if we will copy the file or do a link ?

Reply:
Yes

Reply:
Great, thanks. I will work on it and I will let you know once I finish it :)